### PR TITLE
[proposal] add resource policy plugin

### PR DESCRIPTION
**Pull request by KunWuLuan**
/kind feature

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

**Comment by KunWuLuan** [comment]
@denkensk @Huang-Wei Hi, if you have time, you can help to review the kep, thanks very much. 😆

**Comment by ffromani** [comment]
/ok-to-test

It's a bit of a crazy time for me but I'll try to also add my (non-binding) review

**Comment by Huang-Wei** [comment]
I will try to review it by this week.

**Comment by KunWuLuan** [comment]
Hi, @Huang-Wei , This PR is ready for review. PTAL. Thanks for your time. : )

**Comment by KunWuLuan** [comment]
There is a known issue, when the number of nodes in cluster is larger than 100, scheduler will not find all feasible nodes unless `percentageOfNodesToScore` is set to 100. This can make resource policy lose efficacy.

File: kep/594-resourcepolicy/README.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+## Motivation
+The machines in a Kubernetes cluster are typically heterogeneous, with varying CPU, memory, GPU, and pricing. To efficiently utilize the different resources available in the cluster, users can set priorities for machines of different types and configure resource allocations for different workloads. Additionally, they may choose to delete pods running on low priority nodes instead of high priority ones. 
+
+### Use Cases
+
+1. As a user of cloud services, there are some stable but expensive ECS instances and some unstable but cheaper Spot instances in my cluster. I hope that my workload can be deployed first on stable ECS instances, and during business peak periods, the Pods that are scaled out are deployed on Spot instances. At the end of the business peak, the Pods on Spot instances are prioritized to be scaled in.
+
+### Goals
+
+1. Delvelop a filter plugin to restrict the resource consumption on each unit for different workloads.
```
Comment (ffromani) [inline kep/594-resourcepolicy/README.md:32]: typo: develop

File: kep/594-resourcepolicy/README.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+PreFilter check if the current pods match only one resource policy. If not, PreFilter will reject the pod.
+If yes, PreFilter will get the number of pods on each unit to determine which units are available for the pod
+and write this information into cycleState.
+
+##### Filter
+Filter check if the node belongs to an available unit. If the node doesn't belong to any unit, we will return
+success if the strategy is `prefer`, otherwise we will return unschedulable.
+
+##### Score
+If `priority` and `weight` is set in resource policy, we will schedule pod based on `priority` first. For units with the same `priority`, we will spread pods based on `weight`.
```
Comment (ffromani) [inline kep/594-resourcepolicy/README.md:114]: this may be a silly question, but hey.
So it seems to me that the actual scoring criteria is based on the tuple (priority, weight). Is there any reason to use two values instead of just one single value?
IOW, would it work if we use just `proiority`? What are the pros and cons of using 2 values here?

**Review by ffromani — COMMENTED** [review]
initial pass (sorry for the long delay)
at glance looks sensible, will be asking questions to fully grasp the proposal

File: kep/594-resourcepolicy/README.md (hunk 2 above)
Comment (KunWuLuan) [reply kep/594-resourcepolicy/README.md:114]: I realized that this is a modification that I missed when removing weight earlier. I have already updated the kep

**Comment by KunWuLuan** [comment]
@ffromani @Huang-Wei Hi, do you have any other questions?

**Comment by ffromani** [comment]
The main blocker atm is that the KEP template is not fully filled (e.g.  PostFilter - do we need it? Do we have Graduation criterias? Do we need Production Readiness? [probably not] We probably need a few words in Feature enablement and rollback though)

From what I've read so far I have no major objections. About the architecture and the fitness of this plugin in the existing ecosystem I'd have to defer to someone more experience in the scheduling.

**Comment by KunWuLuan** [comment]
Have update the kep

**Comment by KunWuLuan** [comment]
for #475

**Comment by KunWuLuan** [comment]
@ffromani @Huang-Wei Hi, please have a look when u have time. Thanks for your time. : )

File: kep/594-resourcepolicy/README.md [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    - [Resource Policy Controller](#resource-policy-controller)
+- [Known limitations](#known-limitations)
+- [Test plans](#test-plans)
+- [Graduation criteria](#graduation-criteria)
+- [Feature enablement and rollback](#feature-enablement-and-rollback)
+<!-- /toc -->
+
+## Summary
+This proposal introduces a plugin to allow users to specify the priority of different resources and max resource 
+consumption for workload on differnet resources.
```
Comment (Huang-Wei) [inline kep/594-resourcepolicy/README.md:27]:
```suggestion
This proposal introduces a plugin that enables users to set priorities for various resources and define maximum resource consumption limits for workloads across different resources.
```

File: kep/594-resourcepolicy/README.md [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+### Use Cases
+
+1. As a user of cloud services, there are some stable but expensive ECS instances and some unstable but cheaper Spot 
+instances in my cluster. I hope that my workload can be deployed first on stable ECS instances, and during business peak 
+periods, the Pods that are scaled out are deployed on Spot instances. At the end of the business peak, the Pods on Spot 
+instances are prioritized to be scaled in.
+
+### Goals
+
+1. Develop a filter plugin to restrict the resource consumption on each unit for different workloads.
```
Comment (Huang-Wei) [inline kep/594-resourcepolicy/README.md:44]: could you elaborate on "unit"? do you mean one kind of resource?

File: kep/594-resourcepolicy/README.md [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+2. Develop a score plugin to favor nodes matched by a high priority unit.
+3. Automatically setting deletion costs on Pods to control the scaling in sequence of workloads through a controller.
+
+### Non-Goals
+
+1. Modify the workload controller to support deletion costs. If the workload don't support deletion costs, scaling in 
+sequence will be random.
+2. When creating a ResourcePolicy, if the number of Pods has already violated the quantity constraint of the
```
Comment (Huang-Wei) [inline kep/594-resourcepolicy/README.md:52]: could you reword this to be more readable? basically, just use a positive tone to describe it as a goal that is out of scope of this KEP.

File: kep/594-resourcepolicy/README.md [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        - value2
+  - name: unit3
+    priority: 4
+    maxCount: 20
+    nodeSelector:
+      matchLabels:
+        key1: value3
+```
+
+`Priority` define the priority of each unit. Pods will be scheduled on units with a higher priority.
```
Comment (Huang-Wei) [inline kep/594-resourcepolicy/README.md:108]: is `Prirority` is under `units[*]`? if so, it'd read better to use a bullet to explain it under a top-level `units` API field.

File: kep/594-resourcepolicy/README.md [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    nodeSelector:
+      matchLabels:
+        key1: value3
+```
+
+`Priority` define the priority of each unit. Pods will be scheduled on units with a higher priority. 
+If all units have the same priority, resourcepolicy will only limit the max pod on these units.
+
+`Strategy` indicate how we treat the nodes doesn't match any unit. 
+If strategy is `required`, the pod can only be scheduled on nodes that match the units in resource policy.
```
Comment (Huang-Wei) [inline kep/594-resourcepolicy/README.md:112]: In a multi-tenant cluster, given ResourcePolicy is a namespace-scoped CR, but nodes/node pools might be shared across tenants. Does it mean that if a tenant set a high-priority policy with strategy `required`, it could make other tenants who don't set or set low-priority policy get less chances to get their workloads scheduled?

File: kep/594-resourcepolicy/README.md [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+If strategy is `prefer`, the pod can be scheduled on all nodes, these nodes not match the units will be 
+considered after all nodes match the units. So if the strategy is `required`, we will return `unschedulable` 
+for those nodes not match the units.
+
+`MatchLabelKeys` indicate how we group the pods matched by `podSelector` and `matchPolicy`, its behavior is like 
+`MatchLabelKeys` in `PodTopologySpread`.
+
+`matchPolicy` indicate if we should ignore some kind pods when calculate pods in certain unit.
+
+If `forceMaxNum` is set `true`, we will not try the next units when one unit is not full, this property have no effect
```
Comment (Huang-Wei) [inline kep/594-resourcepolicy/README.md:122]: Could you reword this sentence a bit?

File: kep/594-resourcepolicy/README.md [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+when `max` is not set in units.
+
+### Implementation Details
+
+#### Scheduler Plugins
+
+For each unit, we will record which pods were scheduled on it to prevent too many pods scheduled on it.
+
+##### PreFilter
```
Comment (Huang-Wei) [inline kep/594-resourcepolicy/README.md:131]: Let’s avoid using `#####` for paragraph indentation. The maximum indentation level should be `####` in most cases.

File: kep/594-resourcepolicy/README.md (hunk 6 above)
Comment (Huang-Wei) [reply kep/594-resourcepolicy/README.md:108]: I think we missed a detailed explanation for `units`.

Comment (KunWuLuan) [reply kep/594-resourcepolicy/README.md]: Updated the kep. Thanks

File: kep/594-resourcepolicy/README.md (hunk 8 above)
Comment (KunWuLuan) [reply kep/594-resourcepolicy/README.md]: Rewrote this sentence. Pods will be matched by the ResourcePolicy in same namespace when the `.spec.podSelector`. And if `.spec.matchPolicy.ignoreTerminatingPod` is `true`, pods with Non-Zero `.spec.deletionTimestamp` will be ignored.

File: kep/594-resourcepolicy/README.md (hunk 9 above)
Comment (KunWuLuan) [reply kep/594-resourcepolicy/README.md]: Ok, no problem.

**Comment by Huang-Wei** [comment]
@KunWuLuan I guess the toc needs a refresh. Could you re-run `hack/update-toc.sh`?

**Comment by 13567436138** [comment]
where is implemention？it has been a year ago。

**Comment by KunWuLuan** [comment]
I will give an implementation recently
